niceman/interface/exec.py

```diff
@@ -32,6 +33,211 @@
+
+    def __init__(self, resource, command, cmd_args):
+        self.resource = resource
+        self.session = resource.get_session()
+        self.command = command
+        self.cmd_args = cmd_args
+
+    def pre_command(self):
+        pass
+
+    def execute(self):
+        """Run the full command defined by `command` and `cmd_args`.
+
+        Returns
+        -------
+        A tuple, (stdout, stderr).
+        """
+        raise NotImplementedError
+
+    def post_command(self):
+        pass
+
+    def __call__(self):
+        self.pre_command()
+        try:
+            out, err = self.execute()
+        finally:
+            self.post_command()
+        return out, err
+
+
+class PlainCommand(CommandAdapter):
+
+    def execute(self):
+        return self.session.execute_command([self.command] + self.cmd_args)
+
+
+class InternalCommand(CommandAdapter):
+
+    def execute(self):
+        self.session.niceman_exec(self.command, self.cmd_args)
+        return None, None  # Return same form as Session.execute_command.
+
+
+class TracedCommand(CommandAdapter):
+    """A tracer-wrapped command.
+
+    In the pre-command step, a tracer (extracted from Reprozip) is downloaded
+    to the local machine if not already present.  This tracer is transferred to
+    the resource on which the command will be executed.
+
+    After the command is executed under the tracer, the post-command step
+    downloads the trace artifacts locally and calls `niceman retrace` on the
+    Reprozip spec.
+    """
+
+    def __init__(self, resource, command, cmd_args,
+                 remote_dir=None, local_dir=None):
+        super(TracedCommand, self).__init__(resource, command, cmd_args)
+
+        if not external_versions["reprozip"]:
+            raise RuntimeError("Using --trace requires ReproZip, "
+                               "a Linux-specific dependency")
+
+        self.tracer_md5sum = "d8561c1bc528592b21c0e28d6f32c0a4"
+        # adding two random characters to avoid collisions etc
+        # The id for the execution so we could pick up all the log and trace
+        # files for local storage
+        self.exec_id = "{}-{}".format(
+            time.strftime("%Y%m%d%H%M%S"),
+            str(uuid.uuid4())[:2])
+
+        # Local session variables
+        local_cache_dir = local_dir or op.expanduser('~/.cache/niceman')
+        self.local_tracer_dir = op.join(local_cache_dir,
+                                        "tracers",
+                                        self.tracer_md5sum)
+        self.local_trace_dir = op.join(local_cache_dir, 'traces', self.exec_id)
+        self.local_tracer_gz = op.join(self.local_tracer_dir,
+                                       "niceman_trace.gz")
+
+        # Remote session variables
+        self.remote_dir = remote_dir
+        self.remote_trace_dir = None
+        self.remote_tracer = None
+
+    def _prepare_local(self):
+        if not op.exists(self.local_tracer_gz):
+            import hashlib
```

Comment:
some log msg about the upcoming download might be appreciated

Reply:
good point